src/main/kotlin/com/manimdsl/ASTExecutor.kt

```diff
@@ -0,0 +1,83 @@
+
+import java.util.*
+
+// Wrapper classes for values of variables while executing code
+sealed class ExecValue
+
+data class DoubleValue(val value: Double) : ExecValue()
+
+data class StackValue(val stack: Stack<Double>) : ExecValue()
+
+object EmptyValue : ExecValue()
+
+class ASTExecutor(private val program: ProgramNode) {
+
+    // Map of all the variables and there current values
+    private val variables = mutableMapOf<String, ExecValue>()
+
+    // Keep track of next instruction to execute
+    private var programCounter = 0
+
+    private fun executeExpression(node: ExpressionNode): ExecValue {
+        return when (node) {
+            is IdentifierNode -> variables[node.identifier]!!
+            is NumberNode -> DoubleValue(node.double)
+            is MethodCallNode -> executeMethodCall(node)
+            is ConstructorNode -> executeConstructor(node)
+            is AddExpression -> executeBinaryOp(node) { x, y -> x + y }
+            is SubtractExpression -> executeBinaryOp(node) { x, y -> x - y }
+            is MultiplyExpression -> executeBinaryOp(node) { x, y -> x * y }
+            is PlusExpression -> executeUnaryOp(node) { x -> x }
+            is MinusExpression -> executeUnaryOp(node) { x -> -x }
+        }
+    }
+
+    private fun executeMethodCall(node: MethodCallNode): ExecValue {
+        return when (val ds = variables[node.instanceIdentifier]) {
+            is StackValue -> {
+                return when (node.methodIdentifier) {
+                    "push" -> DoubleValue(ds.stack.push((executeExpression(node.arguments[0]) as DoubleValue).value))
+                    "pop" -> DoubleValue(ds.stack.pop())
+                    else -> EmptyValue
+                }
+            }
+            else -> EmptyValue
+        }
+    }
+
+    private fun executeConstructor(node: ConstructorNode): ExecValue {
+        return when (node.type) {
+            StackType -> StackValue(Stack())
+        }
+    }
+
+    private fun executeUnaryOp(node: UnaryExpression, op: (first: Double) -> Double): ExecValue {
+        return DoubleValue(op((executeExpression(node.expr) as DoubleValue).value))
+    }
+
+    private fun executeBinaryOp(node: BinaryExpression, op: (first: Double, seconds: Double) -> Double): ExecValue {
+        return DoubleValue(
+            op(
+                (executeExpression(node.expr1) as DoubleValue).value,
+                (executeExpression(node.expr2) as DoubleValue).value
+            )
+        )
+    }
+
+    // Returns whether the program is complete and the state of all the variables after executing the statement
+    fun executeNextStatement(): Pair<Boolean, MutableMap<String, ExecValue>> {
+
+        val node = program.statements[programCounter]
+        programCounter++
+        when (node) {
+            is DeclarationNode -> variables[node.identifier] = executeExpression(node.expression)
+            is AssignmentNode -> variables[node.identifier] = executeExpression(node.expression)
```

Comment:
Okay so we will need to interleave the use of the VM and generation of linear IR code. Since intermediate values are not persisted

Reply:
Yes, thats the plan